### PR TITLE
EVG-5994: Fix time taken and makespan in UI

### PR DESCRIPTION
**Pull request by ybrill**
The UI was miscalculating the time taken for builds and the makespan for versions.
Both of these were the result of assumptions about tasks that didn't hold for display tasks since execution tasks run in parallel so `time_taken != finish_time - start_time`.

### CODE REVIEW

**Comment by bsamek** [comment]
I think this broke js-test.

File: public/static/js/build.js [hunk 1]
```diff
@@ -247,13 +248,10 @@ mciModule.controller('BuildViewController', function($scope, $http, $timeout, $r
...
     )
-    $scope.totalTimeMS = _.reduce(
-      _.map(finishedOnly,
-        function(x){return new Date(x.Task.finish_time) - new Date(x.Task.start_time)}),
-        function(sum, el){return sum+el},
-        0)
-    };
 
+    $scope.totalTimeMS = _.reduce(_.pluck(finishedOnly, "Task"), function(x, y){return x.time_taken+y.time_taken}, 0) / nsPerMs;
+    };
```
Comment (bsamek) [inline public/static/js/build.js]: This should be indented to the same level as the scope that it closes.

File: public/static/js/build.js [hunk 2]
```diff
@@ -247,13 +248,10 @@ mciModule.controller('BuildViewController', function($scope, $http, $timeout, $r
...
       }
     )
-    $scope.totalTimeMS = _.reduce(
-      _.map(finishedOnly,
-        function(x){return new Date(x.Task.finish_time) - new Date(x.Task.start_time)}),
-        function(sum, el){return sum+el},
-        0)
-    };
 
+    $scope.totalTimeMS = _.reduce(_.pluck(finishedOnly, "Task"), function(x, y){return x.time_taken+y.time_taken}, 0) / nsPerMs;
```
Comment (bsamek) [inline public/static/js/build.js]: Where they are display tasks, this should get the display tasks xor their execution tasks. How does this handle that?

File: public/static/js/version.js [hunk 3]
```diff
@@ -192,22 +192,14 @@ mciModule.controller('VersionController', function($scope, $rootScope, $location
...
     });
     var taskStartTimes = _.filter(_.pluck(tasks, "start_time").map(function(x){return new Date(x)}), nonZeroTimeFilter).sort(dateSorter);
-    var taskEndTimes = _.filter(tasks.map(function(x){
-        if(x.time_taken == 0 || +new Date(x.start_time) == +new Date(0)){
-            return new Date(0);
-        }else{
-            return new Date((+new Date(x.start_time)) + (x.time_taken/nsPerMs));
-        }
-    }), nonZeroTimeFilter).sort(dateSorter);
+    var taskEndTimes = _.filter(_.pluck(tasks, "finish_time").map(function(x){return new Date(x)}), nonZeroTimeFilter).sort(dateSorter);
```
Comment (bsamek) [inline public/static/js/version.js]: How will this handle finish times which are not yet set?

File: service/version.go [hunk 4]
```diff
@@ -153,24 +153,30 @@ func (uis *UIServer) versionPage(w http.ResponseWriter, r *http.Request) {
...
+			if err != nil {
+				uis.LoggedError(w, r, http.StatusInternalServerError, err)
+				return
+			}
+			if taskFromDb == nil {
+				grip.Error(message.Fields{
 					"task_id": t.Id,
 					"version": projCtx.Version.Id,
 					"request": gimlet.GetRequestID(ctx),
 					"message": "version references task that does not exist",
```
Comment (bsamek) [inline service/version.go]: not your code, but should read "_build_ references..."

File: service/version.go [hunk 5]
```diff
@@ -153,24 +153,30 @@ func (uis *UIServer) versionPage(w http.ResponseWriter, r *http.Request) {
 					DisplayName: t.DisplayName,
 				}}
 
-			// TODO: this loop would probably work better
```
Comment (bsamek) [inline service/version.go]: This changes page loads for versions that don't have many started tasks to make many more db queries. Could this be reduced to a single aggregation per version, or an aggregation per build?

File: public/static/js/build.js (hunk 1 above)
Comment (ybrill) [reply public/static/js/build.js]: done.

File: public/static/js/build.js (hunk 2 above)
Comment (ybrill) [reply public/static/js/build.js]: This is no longer relevant as the calculation of totalTime and makespan was moved to the backend.
But the answer _was_ (:grin:) the only tasks included in the count were those in the build's task cache, which excluded exec tasks.

File: public/static/js/version.js (hunk 3 above)
Comment (ybrill) [reply public/static/js/version.js]: also no longer applicable, but the answer would have been that we filter out tasks with `status != success or failed` as well as zero time

File: service/version.go (hunk 5 above)
Comment (ybrill) [reply service/version.go]: now makes a single indexed find query for the version.

File: service/version.go (hunk 4 above)
Comment (ybrill) [reply service/version.go]: done.

**Pull request merged**